## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [4]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [6]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [8]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [10]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.sample(10)
df.tail()
df.columns
df.dtypes



first           object
last            object
job             object
phone           object
email           object
client_id        int64
order_id         int64
order_date      object
order_week       int64
order_year       int64
item_id         object
category        object
subcategory     object
unit_price     float64
unit_cost      float64
unit_weight    float64
qty              int64
line_number      int64
dtype: object

In [18]:
# What three item categories had the most entries?
most_common_category = df['category'].value_counts().head(3)
most_common_category

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [28]:
# For the category with the most entries, which subcategory had the most entries?
most_common_category = df['category'].value_counts().index[0]
most_common_subcategory = df[df['category'] == most_common_category].reset_index(drop=True)['subcategory'].value_counts().index[0]
most_common_subcategory

'bathroom supplies'

In [30]:
# Which five clients had the most entries in the data?
df['client_id'].value_counts().head(5)

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [34]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head(5).index.tolist()
top_5_clients

[33615, 66037, 46820, 38378, 24741]

In [38]:
# How many total units (the qty column) did the client with the most entries order order?
total_units = df[df['client_id'] == top_5_clients[0]]['qty'].sum()
total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [42]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['line_subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  line_subtotal  
0     1096.80     762.71         7.50  105            1      115164.00  
1       24.95      15.09         1.49   21            0         523.95  
2       13.52       7.86         1.68   39            6         527.28  
3       36.42      24.85         1.23   29            3        1056.18  
4      195.10     108.17        46.43   20            1        3902.00

In [44]:
# Create a column for shipping price.

df['total_weight'] = df['unit_weight'] * df['qty']
df['shipping_price'] = df.apply(lambda row: row['total_weight'] * 7 if row['total_weight'] > 50 else row['total_weight'] * 10, axis=1)

df[['unit_price', 'unit_weight', 'qty', 'total_weight', 'shipping_price']]

unit_price  unit_weight   qty  total_weight  shipping_price
0         1096.80         7.50   105        787.50         5512.50
1           24.95         1.49    21         31.29          312.90
2           13.52         1.68    39         65.52          458.64
3           36.42         1.23    29         35.67          356.70
4          195.10        46.43    20        928.60         6500.20
...           ...          ...   ...           ...             ...
54634       83.13         2.25    33         74.25          519.75
54635      206.59        11.70    47        549.90         3849.30
54636       65.66         4.16   475       1976.00        13832.00
54637        1.48        18.04   112       2020.48        14143.36
54638        3.01         2.07  1031       2134.17        14939.19

[54639 rows x 5 columns]

In [60]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['line_price'] = (df['line_subtotal'] + df['shipping_price']) * 1.0925

df[['line_subtotal', 'shipping_price', 'line_price']].head(3)

line_subtotal  shipping_price     line_price
0      115164.00         5512.50  131839.076250
1         523.95          312.90     914.258625
2         527.28          458.64    1077.117600

In [64]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head(3)

first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   

    line_profit  
0  51702.026250  
1    582.468625  
2    758.817600  

[3 rows x 24 columns]

In [68]:
# Create a column for the profit of each line using line cost and line price
df['line_profit'] = df['line_price'] - df['line_cost']
df.head(3)


first     last              job               phone  \
0   Donald  Harding     Immunologist  793-904-7725x39308   
1  Tiffany    Myers  Music therapist    201.442.4543x942   
2  Shannon   Watson     Immunologist   687.737.9424x8503   

                               email  client_id  order_id  order_date  \
0   harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1  myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2              swatson8146@payne.net      57113   1902144  2023-01-29   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105            1   
1          20        2023  ...     15.09        1.49   21            0   
2           4        2023  ...      7.86        1.68   39            6   

   line_subtotal  total_weight  shipping_price     line_price  line_cost  \
0      115164.00        787.50         5512.50  131839.076250   85597.05   
1         523.95         31.29          312.90     914.258625     629.79   
2         527.28         65.52          458.64    1077.117600     765.18   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  

[3 rows x 24 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [72]:
# Check your work using the totals above
order_ids = [2742071, 2173913, 6128929]
print(df[df['order_id'].isin(order_ids)].groupby('order_id')['line_price'].sum())

order_id
2173913    162388.719300
2742071    152811.897075
6128929    923441.240450
Name: line_price, dtype: float64


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [76]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
top_5_clients = df.groupby('client_id')['qty'].sum().nlargest(5)
print(top_5_clients)
top_5_clients_spend = df[df['client_id'].isin(top_5_clients.index)].groupby('client_id')['line_price'].sum()
print(top_5_clients_spend)

client_id
78965    4033302
86721    1010426
30143     846069
85931     644423
77354     506382
Name: qty, dtype: int64
client_id
30143    3.332885e+07
77354    7.194265e+07
78965    6.559237e+08
85931    2.374170e+08
86721    1.812663e+08
Name: line_price, dtype: float64


In [80]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
# Group the data by client_id and calculate the totals
client_totals = df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'line_price': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
})

# Get the top 5 clients by total units purchased
top_5_clients = client_totals['qty'].nlargest(5).index

# Filter the client_totals DataFrame to only include the top 5 clients
top_5_clients_totals = client_totals.loc[top_5_clients]

# Reset the index to display client_id as a column
top_5_clients_totals = top_5_clients_totals.reset_index()

# Print the result
print(top_5_clients_totals)

   client_id      qty  shipping_price    line_price     line_cost  \
0      78965  4033302      5340756.88  6.559237e+08  4.085708e+08   
1      86721  1010426     64335727.72  1.812663e+08  1.382916e+08   
2      30143   846069      8981905.24  3.332885e+07  2.207846e+07   
3      85931   644423     10122549.19  2.374170e+08  1.281229e+08   
4      77354   506382     23814685.86  7.194265e+07  5.951039e+07   

    line_profit  
0  2.473528e+08  
1  4.297470e+07  
2  1.125039e+07  
3  1.092941e+08  
4  1.243226e+07  


In [104]:
# Format the data and rename the columns to names suitable for presentation.
# Group the data by client_id and calculate the totals
client_totals = df.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum',
    'line_price': 'sum',
    'line_cost': 'sum',
    'line_profit': 'sum'
})

# Get the top 5 clients by total units purchased
top_5_clients = client_totals['qty'].nlargest(5).index

# Filter the client_totals DataFrame to only include the top 5 clients
top_5_clients_totals = client_totals.loc[top_5_clients]

# Reset the index to display client_id as a column
top_5_clients_totals = top_5_clients_totals.reset_index()

# Rename the columns
top_5_clients_totals.columns = [
    'Client ID', 
    'Total Units Purchased', 
    'Total Shipping Cost', 
    'Total Revenue', 
    'Total Cost', 
    'Total Profit'
]

# Format the data
top_5_clients_totals[['Total Shipping Cost', 'Total Revenue', 'Total Cost', 'Total Profit']] = top_5_clients_totals[['Total Shipping Cost', 'Total Revenue', 'Total Cost', 'Total Profit']].round(2)

# Print the result
print(top_5_clients_totals)
# Define the money columns. 

money_columns = ['Total Shipping Cost', 'Total Revenue', 'Total Cost', 'Total Profit']
# Define a function that converts a dollar amount to millions.
def dollars_to_millions(x):
    return x / 1e6

# Apply the currency_format_millions function to only the money columns.
top_5_clients_totals[money_columns] = top_5_clients_totals[money_columns].apply(dollars_to_millions)

# Rename the columns to reflect the change in the money format.
top_5_clients_totals.columns = [
    'Client ID', 
    'Total Units Purchased', 
    'Total Shipping Cost (M)', 
    'Total Revenue (M)', 
    'Total Cost (M)', 
    'Total Profit (M)'
]

top_5_clients_totals

   Client ID  Total Units Purchased  Total Shipping Cost  Total Revenue  \
0      78965                4033302           5340756.88   6.559237e+08   
1      86721                1010426          64335727.72   1.812663e+08   
2      30143                 846069           8981905.24   3.332885e+07   
3      85931                 644423          10122549.19   2.374170e+08   
4      77354                 506382          23814685.86   7.194265e+07   

     Total Cost  Total Profit  
0  4.085708e+08  2.473528e+08  
1  1.382916e+08  4.297470e+07  
2  2.207846e+07  1.125039e+07  
3  1.281229e+08  1.092941e+08  
4  5.951039e+07  1.243226e+07  


Client ID  Total Units Purchased  Total Shipping Cost (M)  \
0      78965                4033302                 5.340757   
1      86721                1010426                64.335728   
2      30143                 846069                 8.981905   
3      85931                 644423                10.122549   
4      77354                 506382                23.814686   

   Total Revenue (M)  Total Cost (M)  Total Profit (M)  
0         655.923670      408.570828        247.352842  
1         181.266297      138.291602         42.974695  
2          33.328850       22.078464         11.250386  
3         237.417001      128.122873        109.294129  
4          71.942652       59.510392         12.432260

In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.


Client ID   Units  Shipping (millions)  Total Revenue (millions)  \
4      24741  239862             5.126448                 82.268892   
3      38378   73667             3.429455                 12.906551   
1      66037   43018             1.395152                 10.259515   
2      46820   75768             1.601449                  9.743794   
0      33615   64313             1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [106]:
sorted_data = top_5_clients_totals.sort_values(by='Total Profit (M)', ascending=False)
sorted_data

Client ID  Total Units Purchased  Total Shipping Cost (M)  \
0      78965                4033302                 5.340757   
3      85931                 644423                10.122549   
1      86721                1010426                64.335728   
4      77354                 506382                23.814686   
2      30143                 846069                 8.981905   

   Total Revenue (M)  Total Cost (M)  Total Profit (M)  
0         655.923670      408.570828        247.352842  
3         237.417001      128.122873        109.294129  
1         181.266297      138.291602         42.974695  
4          71.942652       59.510392         12.432260  
2          33.328850       22.078464         11.250386

In [ ]:
# Top 5 clients by total profit, with respective metrics in millions, 
# sorted in descending order by total profit for targeted strategy development
# This analysis enables businesses to prioritize and tailor their strategies 
# to their most valuable customers, maximizing revenue and growth opportunities